src/zarr/core/array.py

```diff
@@ -188,6 +217,139 @@ def __init__(
         object.__setattr__(self, "order", order_parsed)
         object.__setattr__(self, "codec_pipeline", create_codec_pipeline(metadata=metadata_parsed))
 
+    @overload
```

Comment:
I think this will work for now. We might want to see if using a [TypeGuard](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeguard) in the implementation removes the need for the overload.

I'm not sure how `zarr_format=None` will complicate things, since by definition that relies on a runtime value.